### PR TITLE
[FLINK-14484] Enable to control memory usage of RocksDB via Cache and WriteBufferManager

### PR DESCRIPTION
**Pull request by Myasuka**
## What is the purpose of the change

Enable to control memory usage of RocksDB via Cache and WriteBufferManager.
We would lazy initialize a `LRUCache` and `WriteBufferManager` (which cost memory from the cache) when first keyed state backend created, and share these objects among other RocksDB instances per slot. By doing this, we could limit the memory usage of block cache and write buffers.
We would also set flag `setCacheIndexAndFilterBlocks` as true to limit the memory usage of index&filter. Moreover, we would also set flags `setCacheIndexAndFilterBlocksWithHighPriority` and `setPinL0FilterAndIndexBlocksInCache` as true to minimize the performance impact when we try to limit the total memory (please refer to [cache-index-filter-doc](https://github.com/facebook/rocksdb/wiki/Block-Cache#caching-index-filter-and-compression-dictionary-blocks) for more details)

## Brief change log

  - Introduce a deque of `stateBackendSharedObjects` within `MemoryManager`, which contains objects shared among RocksDB instances per slot.
  - Introduce `state.backend.rocksdb.per-slot.total.memory`, `state.backend.rocksdb.write-buffer.ratio` and `state.backend.rocksdb.high-pri-pool.ratio` to `RocksDBOptions`.
  - If total memory is limited, RocksDB would share cache and writebuffer manager to limit the total memory usage.

## Verifying this change

This change added tests and can be verified as follows:
  - Added a new `RocksDBStateBackendBoundedMemoryTest` to verify memory bounded for rocksDB state backend.
  - Extended `RocksDBStateBackendConfigTest` to verify memory control related parameters.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): **yes**, memory control would impact the performance of RocksDB
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs in FLINK-14495

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java [hunk 1]
```diff
@@ -81,6 +84,12 @@
 	/** Reserved memory per memory owner. */
 	private final Map<Object, Map<MemoryType, Long>> reservedMemory;
 
+	/** Reserved closeable shared objects for state backends fetched from this memory manager. */
+	private final AtomicReference<Deque<AutoCloseable>> stateBackendSharedObjects;
```
Comment (carp84) [inline flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java]: I suggest to introduce a `RocksDBSharedObject` wrapping up `Cache` and `WriteBufferManager` objects to attach here to simplify the logic. Please refer to [this review branch](https://github.com/carp84/flink/tree/flink-14484-review) for detailed idea.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBOptions.java [hunk 2]
```diff
@@ -86,4 +86,24 @@
...
+		.key("state.backend.rocksdb.per-slot.total.memory")
+		.stringType()
+		.noDefaultValue()
+		.withDescription("The total memory size shared among all RocksDB instances per slot. " +
+			"This option has no default value which means no bounded memory limit.");
+
+	public static final ConfigOption<Double> WRITE_BUFFER_RATIO = ConfigOptions
+		.key("state.backend.rocksdb.write-buffer.ratio")
+		.doubleType()
+		.defaultValue(0.8)
```
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBOptions.java]: Why do we set this default value to `0.8`? Maybe `0.5` (read:write=1:1) is a better idea?

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendBoundedMemoryTest.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+			Assert.assertTrue(sharedObjects.getLast() instanceof LRUCache);
+			Assert.assertEquals(2, sharedObjects.size());
+			Assert.assertEquals(lruCache, sharedObjects.getLast());
+			Assert.assertEquals(writeBufferManager, sharedObjects.getFirst());
+		} finally {
+			keyedStateBackend1.close();
+			keyedStateBackend2.close();
+			// even keyed state backend closed, cache and write buffer manager would not be disposed.
+			Assert.assertTrue(lruCache.isOwningHandle());
+			Assert.assertTrue(writeBufferManager.isOwningHandle());
```
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendBoundedMemoryTest.java]: Extract this into a separate test case.

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendBoundedMemoryTest.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	}
+
+	@After
+	public void shutDown() {
+		this.memoryManager.shutdown();
+		Assert.assertTrue(memoryManager.getStateBackendSharedObjects().get().isEmpty());
+	}
+
+	@Test
+	public void testSharedObjectsInitializeOnlyOnce() throws IOException {
```
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendBoundedMemoryTest.java]: Suggest to add another test case for the race condition case under concurrency.

File: flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java (hunk 1 above)
Comment (Myasuka) [reply flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java]: Thanks for this idea.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBOptions.java (hunk 2 above)
Comment (Myasuka) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBOptions.java]: Sounds reasonable.

**Pull request closed without merging**